enclave/ecalls.cpp

```diff
@@ -41,59 +41,66 @@ int enclave_XGDMatrixCreateFromEncryptedFile(const char *fname, int silent, DMat
   return XGDMatrixCreateFromEncryptedFile(fname, silent, out);
 }
 
-int enclave_XGBoosterCreate(const DMatrixHandle dmats[], bst_ulong len, BoosterHandle* out) {
+int enclave_XGBoosterCreate(DMatrixHandle dmat_handles[], size_t handle_lengths[], bst_ulong len, BoosterHandle* out) {
   LOG(DEBUG) << "Ecall: XGBoosterCreate";
+
+  // Validate buffers and copy to enclave memory
+  DMatrixHandle dmats[len];
   for (int i = 0; i < len; i++) {
-      check_enclave_ptr(dmats[i]);
+    char* name = dmat_handles[i];
+    size_t nlen = handle_lengths[i];
+    check_host_buffer(name, nlen);
+    dmats[i] = strndup(name, nlen);
```

Comment:
Here and in other ecalls that use `strndup` - do we need to free the returned pointer? Will there be [memory leakage](https://en.cppreference.com/w/c/experimental/dynamic/strndup) if we don't?

Reply:
Good catch! Fixed.